Remove process group barrier and all_reduce function calls from tensorpipe agent

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66708 Remove ProcessGroup from TensorPipeAgent initialization
* #66257 Formatting changes
* **#65946 Remove process group barrier and all_reduce function calls from tensorpipe agent**

Add new function in agent_utils to perform a synchronization of active call counts using store. This is intended to replace the barrier and all_reduce used by the process group in RPC shutdown.

`test_ddp_comparison` and `test_ddp_comparison_uneven_inputs` initialize process groups using the same file_name as RPC agent. At the end the test, the processes call `dist.destroy_process_group()` which clears `_pg_map` and causes FileStore to be garbage collected. In the FileStore destructor, it removes the file that was used. Then when RPC agent is trying to shutdown, it is recreating that file (still unsure about this?) and thus not able to complete barrier.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

Differential Revision: [D31762736](https://our.internmc.facebook.com/intern/diff/D31762736)

H-Huang committed:
Remove process group barrier and all_reduc function calls from tensorpipe agent

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/agent_utils.cpp:
```diff
@@ -40,6 +40,93 @@ std::unordered_map<std::string, worker_id_t> collectNames(
+      desiredValue =
+          createKeyValue(processNum + 1, totalActiveCalls + activeCalls);
+    }
+    newKeyValue = store.compareSet(barrierKey, currKeyValue, desiredValue);
```
Would I be correct if I assume `compareSet` will create the key if it does not exist?

mrshenli commented on torch/csrc/distributed/rpc/agent_utils.cpp:
```diff
@@ -40,6 +40,93 @@ std::unordered_map<std::string, worker_id_t> collectNames(
+      currKeyValue = std::string(valueArr.begin(), valueArr.end());
+      std::tie(processNum, totalActiveCalls) = parseKeyValue(currKeyValue);
+      desiredValue =
+          createKeyValue(processNum + 1, totalActiveCalls + activeCalls);
```
I guess this part is the race condition as it's possible that multiple processes are reading the same `totalActiveCalls` and then compareSet won't sum the together?

mrshenli replied:
I wonder if we should use `add()` instead?

mrshenli replied:
just wanna share a bit more about the idea, assume `add()` is atomic. We can have two keys here, one for the number reported processes, and another for the number of total active calls. Each Process always first update the total active calls, and then update the total number of reported processes. The last process (i.e., the one that sees the total number of reported processes equals the world size) is responsible to set the ready key and every process should then read the total number of active calls on their own. I assume this can also avoid using the while loop? There might be other ways to further simplify it.

H-Huang committed:
Update on "WIP: Remove process group barrier and all_reduce function calls from tensorpipe agent"


WIP: Race condition in new barrier function since there isn't a way to validate that the value written through compare_set was written due to the current process or a different process.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

H-Huang committed:
Update on "WIP: Remove process group barrier and all_reduce function calls from tensorpipe agent"


WIP: Race condition in new barrier function since there isn't a way to validate that the value written through compare_set was written due to the current process or a different process.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

H-Huang committed:
Update on "WIP: Remove process group barrier and all_reduce function calls from tensorpipe agent"


WIP: Race condition in new barrier function since there isn't a way to validate that the value written through compare_set was written due to the current process or a different process.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

H-Huang committed:
Update on "Remove process group barrier and all_reduce function calls from tensorpipe agent"


Add new function in agent_utils to perform a synchronization of active call counts using store. This is intended to replace the barrier and all_reduce used by the process group in RPC shutdown.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

H-Huang committed:
Update on "Remove process group barrier and all_reduce function calls from tensorpipe agent"


Add new function in agent_utils to perform a synchronization of active call counts using store. This is intended to replace the barrier and all_reduce used by the process group in RPC shutdown.

`test_ddp_comparison` and `test_ddp_comparison_uneven_inputs` test fail with these changes. It seems like the RPC agents are not accessing the same store, so the total count of processes never reaches the world size to exit the barrier, still ened to investigate why it is like this only for these test cases. Setting clean_shutdown to false ignores this code path which allows the test to pass.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

mrshenli reviewed (commented):
> test_ddp_comparison and test_ddp_comparison_uneven_inputs test fail with these changes. It seems like the RPC agents are not accessing the same store, so the total count of processes never reaches the world size to exit the barrier, still ened to investigate why it is like this only for these test cases.

Would I be correct that this is addressed?

mrshenli reviewed (commented):
Thanks for the update! The logic in `syncCallCount` looks correct to me. The only thing left is investigating why the two tests failed.

H-Huang committed:
Update on "Remove process group barrier and all_reduce function calls from tensorpipe agent"


Add new function in agent_utils to perform a synchronization of active call counts using store. This is intended to replace the barrier and all_reduce used by the process group in RPC shutdown.

`test_ddp_comparison` and `test_ddp_comparison_uneven_inputs` initialize process groups using the same file_name as RPC agent. At the end the test, the processes call `dist.destroy_process_group()` which clears `_pg_map` and causes FileStore to be garbage collected. In the FileStore destructor, it removes the file that was used. Then when RPC agent is trying to shutdown, it is recreating that file (still unsure about this?) and thus not able to complete barrier.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

Differential Revision: [D31762736](https://our.internmc.facebook.com/intern/diff/D31762736)

[ghstack-poisoned]

H-Huang committed:
Update on "Remove process group barrier and all_reduce function calls from tensorpipe agent"


Add new function in agent_utils to perform a synchronization of active call counts using store. This is intended to replace the barrier and all_reduce used by the process group in RPC shutdown.

`test_ddp_comparison` and `test_ddp_comparison_uneven_inputs` initialize process groups using the same file_name as RPC agent. At the end the test, the processes call `dist.destroy_process_group()` which clears `_pg_map` and causes FileStore to be garbage collected. In the FileStore destructor, it removes the file that was used. Then when RPC agent is trying to shutdown, it is recreating that file (still unsure about this?) and thus not able to complete barrier.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

Differential Revision: [D31762736](https://our.internmc.facebook.com/intern/diff/D31762736)

[ghstack-poisoned]

H-Huang committed:
Update on "Remove process group barrier and all_reduce function calls from tensorpipe agent"


Add new function in agent_utils to perform a synchronization of active call counts using store. This is intended to replace the barrier and all_reduce used by the process group in RPC shutdown.

`test_ddp_comparison` and `test_ddp_comparison_uneven_inputs` initialize process groups using the same file_name as RPC agent. At the end the test, the processes call `dist.destroy_process_group()` which clears `_pg_map` and causes FileStore to be garbage collected. In the FileStore destructor, it removes the file that was used. Then when RPC agent is trying to shutdown, it is recreating that file (still unsure about this?) and thus not able to complete barrier.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

Differential Revision: [D31762736](https://our.internmc.facebook.com/intern/diff/D31762736)

[ghstack-poisoned]

mrshenli reviewed (commented):
For the following failure, can you rebase to https://github.com/pytorch/pytorch/pull/67204?

Please rebase to >= #67204 and  < #67264. The latter broke a bunch of distributed tests. We are still fixing those.

```
======================================================================
ERROR [2.014s]: test_collectives (__main__.PythonProcessGroupTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 457, in wrapper
    self._join_processes(fn)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 676, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 721, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 1 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 575, in run_test
    getattr(self, test_name)()
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 459, in wrapper
    fn()
  File "/var/lib/jenkins/workspace/test/distributed/test_c10d_common.py", line 787, in test_collectives
    dist.init_process_group("dummy", rank=self.rank, world_size=self.world_size)
  File "/opt/conda/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 617, in init_process_group
    _store_based_barrier(rank, store, timeout)
  File "/opt/conda/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 240, in _store_based_barrier
    worker_count = store.add(store_key, 0)
RuntimeError: Broken pipe
```

H-Huang committed:
Update on "Remove process group barrier and all_reduce function calls from tensorpipe agent"


Add new function in agent_utils to perform a synchronization of active call counts using store. This is intended to replace the barrier and all_reduce used by the process group in RPC shutdown.

`test_ddp_comparison` and `test_ddp_comparison_uneven_inputs` initialize process groups using the same file_name as RPC agent. At the end the test, the processes call `dist.destroy_process_group()` which clears `_pg_map` and causes FileStore to be garbage collected. In the FileStore destructor, it removes the file that was used. Then when RPC agent is trying to shutdown, it is recreating that file (still unsure about this?) and thus not able to complete barrier.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

Differential Revision: [D31762736](https://our.internmc.facebook.com/intern/diff/D31762736)

[ghstack-poisoned]

mrshenli reviewed (approved):
stamp to unblock, please wait for this test, and check that  `test_ddp_comparison` and `test_ddp_comparison_uneven_inputs` indeed run there https://github.com/pytorch/pytorch/runs/4024497010?check_suite_focus=true
